platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/http/HttpRequest.java

```diff
@@ -0,0 +1,37 @@
+package com.mapbox.mapboxsdk.http;
+
+import android.support.annotation.Nullable;
+
+public abstract class HttpRequest {
+
+  private static HttpRequest httpRequest;
+
+  static final int CONNECTION_ERROR = 0;
+  static final int TEMPORARY_ERROR = 1;
+  static final int PERMANENT_ERROR = 2;
+
+  public static synchronized void setHttpRequest(@Nullable HttpRequest requestImpl) {
+    httpRequest = requestImpl;
```

Comment:
I think we need to invoke a `cancelAll` method here because we will lose the reference to the calls when the request impl is changed.

Reply:
The idea of this PR is to be able exclude okhttp, a `cancelAll` invocation in this code path on the `OkHttpRequest` will result in the initiating `OkHttpRequest` and the classloader will hit a `NoClassDefFoundError`.

imo switching out httpclients on the fly isn't a supported use-case. You either use the default okhttp implementation or you provide an alternate implementation before requesting any resources.

I'm down with different suggestions to tackle this but don't think there is much we can do other than documenting this clearly in the javadoc.

Reply:
Right, let's document it extensively then.

What about throwing an exception when resources were already requested and a user wants to switch the implementation?

Reply:
👍